### PR TITLE
selection.modify to respect ShadowRoot

### PR DESCRIPTION
**Pull request by zurfyx**
Tables had some logic to override the deletion logic that was incomplete. This prevented new lines from being deleted.

However, when attempting to fix this (and switching the DELETE_CHARACTER command to be unhandled) it uncovered a bigger problem, the fact that selection can expand beyond a ShadowRoot and cause cells to be deleted. This shouldn't be the responsibility of tables, so I'm handling the ShadowRoot case appropriately in Selection but rather generic.

Closes https://github.com/facebook/lexical/issues/3278

### CODE REVIEW

File: packages/lexical/src/LexicalNode.ts [hunk 1]
```diff
@@ -590,6 +590,20 @@ export class LexicalNode {
     return this.getTextContent().length;
   }
 
+  /**
+   * Returns the nearest ShadowRoot or RootNode
+   */
+  getRoot(): RootNode | ElementNode {
```
Comment (acywatson) [inline packages/lexical/src/LexicalNode.ts:596]: getRootOrShadowRoot?

Sounds clunky but not sure about making this the same name (without $) as the method that always returns the true root.

File: packages/lexical/src/LexicalSelection.ts [hunk 2]
```diff
@@ -1768,7 +1766,7 @@ export class RangeSelection implements BaseSelection {
     // from getTargetRanges(), and is also better than trying to do it ourselves
     // using Intl.Segmenter or other workarounds that struggle with word segments
     // and line segments (especially with word wrapping and non-Roman languages).
-    $moveNativeSelection(
+    moveNativeSelection(
```
Comment (acywatson) [inline packages/lexical/src/LexicalSelection.ts:1769]: y? Looks like we use $ methods inside

Comment (zurfyx) [reply packages/lexical/src/LexicalSelection.ts:1769]: Hm, we don't

**Review by acywatson — APPROVED** [review]

File: packages/lexical/src/LexicalNode.ts (hunk 1 above)
Comment (zurfyx) [reply packages/lexical/src/LexicalNode.ts:596]: > TBH, wouldn't add LexicalNode#getRoot() and keep it as a util method, as our node classes already bloated -- @fantactuka 

Makes sense, moved away from LexicalNode

**Review by tylerjbainbridge — APPROVED** [review]

**Pull request merged**